pmsx003: add standard particle, particle counts

warthog9 opened the pull request:
# What does this implement/fix? 

These two commits add support for
(1) Decilitres
(2) Which are then in turn used to add the rest of the sensor data from the PMSx003 series of sensors which return more data than the standard PM1.0, PM2.5, and PM10 sensor data.  This new data includes:

- PM1.0, PM2.5, PM10 under standard pressure (new)
- PM1.0, PM2.5, PM10 under atmospheric pressure (this is existing an unchanged functionality)
- Particle count per decilitre of particles sized 0.3um, 0.5um, 1.0um, 2.5um, 5.0um, and 10.0um

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):**

This doesn't fix an existing bug, though it adds functionality to an existing sensor

Bug open about this now (reported after initial pull submitted)
esphome/feature-requests/issues/1201

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**
esphome/esphome-docs#1102
  
# Test Environment

- [X] ESP32
- [ ] ESP8266
- [ ] Windows
- [ ] Mac OS
- [X] Linux

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
- platform: pmsx003
   type: PMSX003
   uart_id: bus_uart_pms
   pm_1_0_std:
     name: "Particulate Matter <1.0µm Concentration"
     id: pm_1_0_std
     filters:
       - throttle: 10s
   pm_2_5_std:
     name: "Particulate Matter <2.5µm Concentration"
     id: pm_2_5_std
     filters:
       - throttle: 10s
   pm_10_0_std:
     name: "Particulate Matter <10.0µm Concentration"
     id: pm_10_0_std
     filters:
       - throttle: 10s
   pm_0_3um:
     name: "Particulate Count >0.3um"
     id: pm_0_3um
     filters:
       #- throttle: 10s
       #- debounce: 1s
       #- delta: 1
       - sliding_window_moving_average:
           window_size: 60
           send_every: 45

```

# Explain your changes

The PMSx003 sensors send a lot more data than just the 3 values esphome currently reads out, including the PM values under standard particles, and the count of particles above certain size thresholds.  This adds a fair amount more monitoring data that can be easily tracked (particularly particle size counts) that are useful to getting an idea of what's going on in the greater environment.  This basically extends the existing data out to include the data that's already being fetched across the uart.

## Checklist:
  - [X] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

warthog9 committed:
Add Decilitre constant

This adds the decilitre constant, or 1/10th of a litre of volume

Signed-off-by: John 'Warthog9' Hawley <warthog9@eaglescrag.net>

warthog9 committed:
components/pmsx003: Add support for missing values from sensor

The PMSx003 sensors not only return AQI indicators for PM1, PM2.5,
and PM10 but also the standar PMs (in addition to the currently reported
values which are the values under atmospheric pressure, as well as the
particle counts by size.  The later values being in decilitres.  This adds
in the rest of those data points from the sensor for tracking by
esphome.

Signed-off-by: John 'Warthog9' Hawley <warthog9@eaglescrag.net>

glmnet commented on esphome/const.py:
```diff
@@ -645,6 +654,7 @@
 
 UNIT_AMPERE = "A"
 UNIT_CELSIUS = "°C"
+UNIT_COUNT_DECILITRE = "/dL"
```
why is the unit starting with /?

warthog9 replied:
At least in this case it's a count per decilitre, trying to make it more obvious that it's a <count> per unit volume as I'm unaware of a good way to denote that otherwise.  Happy to change that if there's a better way to represent that, I just don't want someone to misconstrue the number as 1000 dL where that may be a very different things thant 1000 /dL

jesserockz replied:
Maybe this could follow the line below `UNIT_COUNTS_PER_CUBIC_METER = "#/m³"` and use the `#` to represent count

warthog9 replied:
the problem with putting "#" in front, I found, it is messes up things like Grafana so that the measurement doesn't actually show up, so if you are tracking information along the way it can sometimes get miss-interpretted

glmnet commented on tests/test3.yaml:
```diff
@@ -395,6 +395,24 @@ sensor:
+    pm_10_0_std:
+      name: 'PM 10.0 Standard Atmospher Concentration'
+    pm_0_3um:
+      name: 'Particulate Count >0.3um'
```
I know this is test file, so this is a question, is the > means count of particles greater than 0.3um? or is an arrow, because all detectable particles would be > 0.3um

warthog9 replied:
it would be a count of particles larger than 0.3um.  Since the sensor reports 0.3, 0.4, 1.0, 2.5, 5.0, 10 larger sizes will be, obviously, included in smaller size counts.  For instance a device on my desk is reporting:
>0.3um: 12
>0.5um: 6
>1.0um: 0 (and subsequent sizes also 0 obviously)

the >0.3um would include particles 0.3um and larger, which would also be the 6 from >0.5um.  So about 6 particles between 0.3um and 0.5um, and 6 particles between 0.5um and 1.0um were present while I was writing this

glmnet reviewed (commented):
Just two inline questions

genestealer commented:
I have just been using the PMSX5003ST so am particularly interested in seeing this get merged.

Have you considered also adding in the concentration units (environmental), in addition to the particle size binning, like adafruit did? https://github.com/adafruit/Adafruit_PM25AQI

warthog9 commented:
> I have just been using the PMSX5003ST so am particularly interested in seeing this get merged. [esphome/feature-requests#1201](https://github.com/esphome/feature-requests/issues/1201)
> 
> Have you considered also adding in the concentration units (environmental), in addition to the particle size binning, like adafruit did? https://github.com/adafruit/Adafruit_PM25AQI
> 
> Hold on a sec, looks as if it already is doing environmental, as I see someone else has a pull request open for adding in the Standard CF1 Readings #1284 so it looks like I may have misunderstood which values are being returned by default.

This is already taken care of in the existing code (which returns the environmental atmosphere version) and the _std version of the same (standard atmospheric version).  I specifically didn't change the existing type as to not break existing usage

warthog9 committed:
components/sensor/pmsx003: merge PMS5003 w/ PMS5003ST assignment

The PMS5003 and PMS5003ST differ only in the last couple of bytes
acquired off the stream which represent the temperature, humidity,
and voc/formaldehyde.  This basically removes duplicated code, and
pulls the ST specific bits first, then processes the rest of it
as if it was a PMS5003.

Signed-off-by: John 'Warthog9' Hawley <warthog9@eaglescrag.net>

warthog9 commented:
Ok with respect to the PMS5003ST, I've got an additional patch in my git repo that merges the PMS5003 and PMS5003ST processing, by pulling the temperature, humidity, and voc first then letting the case fall through to the PMSx003 case and processing the rest there.  Since I don't have a PMS5003ST I'd love it if someone who has one could test it, but I'll hold that patch off for now till someone can test it and make sure it works

jesserockz commented on esphome/const.py:
```diff
@@ -692,6 +701,8 @@
 UNIT_CELSIUS = "°C"
 UNIT_COUNTS_PER_CUBIC_METER = "#/m³"
 UNIT_CUBIC_METER = "m³"
+UNIT_COUNT_DECILITRE = "/dL"
+UNIT_COUNTS_PER_CUBIC_METER = "#/m³"
```
You re-added the `UNIT_COUNTS_PER_CUBIC_METER` and also `UNIT_COUNT_DECILITRE` is out of order.

warthog9 replied:
Sorry, bad rebase, should be fixed and rebased up now

genestealer commented:
> Ok with respect to the PMS5003ST, I've got an additional patch in my git repo that merges the PMS5003 and PMS5003ST processing, by pulling the temperature, humidity, and voc first then letting the case fall through to the PMSx003 case and processing the rest there. Since I don't have a PMS5003ST I'd love it if someone who has one could test it, but I'll hold that patch off for now till someone can test it and make sure it works

I have the PMS5003ST if you want me to test your code?

warthog9 commented:
> > Ok with respect to the PMS5003ST, I've got an additional patch in my git repo that merges the PMS5003 and PMS5003ST processing, by pulling the temperature, humidity, and voc first then letting the case fall through to the PMSx003 case and processing the rest there. Since I don't have a PMS5003ST I'd love it if someone who has one could test it, but I'll hold that patch off for now till someone can test it and make sure it works
> 
> I have the PMS5003ST if you want me to test your code?

That would be awesome, https://github.com/warthog9/esphome/tree/pms5003st-merge has the 5003 + 5003ST merged into a single "thing"

WhatIsMyOldUserName commented:
What is still required to get this merged? These are excellent features that were missing.

warthog9 commented:
> What is still required to get this merged? These are excellent features that were missing.

probably 3 things:
1) It would be nice if someone can test the pms5003st-merge branch above, I don't have a 5003st, so can't confirm that works.
2) I likely need to rebase this again, the last time I got things up to date was in June, and I'm sure it needs some minor adjustments
3) The CI to be run, but since I'm a new contributors (and as a measure by Github to prevent mining operations) someone needs to approve the CI run, and assuming that passes merge it.

genestealer commented:
I have a pms5003st, my apologies though what I would need to do in order to test the branch, can you offer any guidance? I'm currently running esphome within home assistant io

warthog9 commented:
@genestealer basically https://github.com/warthog9/esphome/tree/pms5003st-merge has the PMS5003 and the PMS5003ST "merged" (basically the same code path reading both sensors instead of it being fully separate).  Basically if you can compile / load that on a device and read all the data out of the PMS5003ST, then it works and should be fine

warthog9 commented:
@genestealer ok just rebased, so that knocks one of the todo items off the list

OttoWinter commented on esphome/const.py:
```diff
@@ -437,10 +437,19 @@
 CONF_PIXEL_MAPPER = "pixel_mapper"
 CONF_PLATFORM = "platform"
 CONF_PLATFORMIO_OPTIONS = "platformio_options"
+CONF_PM_0_3um = "pm_0_3um"
+CONF_PM_0_5um = "pm_0_5um"
 CONF_PM_1_0 = "pm_1_0"
+CONF_PM_1_0_STD = "pm_1_0_std"
+CONF_PM_1_0um = "pm_1_0um"
 CONF_PM_10_0 = "pm_10_0"
+CONF_PM_10_0_STD = "pm_10_0_std"
+CONF_PM_10_0um = "pm_10_0um"
 CONF_PM_2_5 = "pm_2_5"
+CONF_PM_2_5_STD = "pm_2_5_std"
+CONF_PM_2_5um = "pm_2_5um"
 CONF_PM_4_0 = "pm_4_0"
+CONF_PM_5_0um = "pm_5_0um"
```
```suggestion
CONF_PM_0_3UM = "pm_0_3um"
CONF_PM_0_5UM = "pm_0_5um"
CONF_PM_1_0 = "pm_1_0"
CONF_PM_1_0_STD = "pm_1_0_std"
CONF_PM_1_0UM = "pm_1_0um"
CONF_PM_10_0 = "pm_10_0"
CONF_PM_10_0_STD = "pm_10_0_std"
CONF_PM_10_0UM = "pm_10_0um"
CONF_PM_2_5 = "pm_2_5"
CONF_PM_2_5_STD = "pm_2_5_std"
CONF_PM_2_5UM = "pm_2_5um"
CONF_PM_4_0 = "pm_4_0"
CONF_PM_5_0UM = "pm_5_0um"
```

All keys in `conf.py` are uppercase (because they're constants)

warthog9 replied:
updated

OttoWinter reviewed (approved):
:tada: Thanks!
